File: 36_集約/go/main.go
package main

import (
	"fmt"
	"math"
	"strconv"
)

func validateNumberOrDone(input string) (bool, string) {
	if input == "done" {
		return true, input
	}
	_, err := strconv.Atoi(input)
	if err != nil {
		return false, ""
	}
	return true, input
}
func dispNums(nums []int) {
	fmt.Print("Numbers: ")
	sep := ""
	for _, n := range nums {
		fmt.Printf("%s%d", sep, n)
		sep = ", "
	}
	fmt.Println("")
}
func maximum(nums []int) int {
	if len(nums) == 0 {
		panic("size must not be zero.")
	}
	max := nums[0]
	for i := 0; i < len(nums); i++ {
		if nums[i] > max {
			max = nums[i]
		}
	}
	return max
}
func minimum(nums []int) int {
	if len(nums) == 0 {
		panic("size must not be zero.")
	}
	min := nums[0]
	for i := 0; i < len(nums); i++ {
		if nums[i] < min {
			min = nums[i]
		}
	}
	return min
}
func average(nums []int) float64 {
	if len(nums) == 0 {
		panic("size must not be zero.")
	}
	sum := 0.0
	for _, n := range nums {
		sum += float64(n)
	}
	return sum / float64(len(nums))
}

func standardDiviation(nums []int) float64 {
	if len(nums) == 0 {
		panic("size must not be zero.")
	}

	average := average(nums)
	tmpSum := 0.0
	for _, e := range nums {
		diff := average - float64(e)
		tmpSum += (diff * diff)
	}
	v := tmpSum / float64(len(nums))
	return math.Sqrt(v)
}

func main() {
	nums := []int{}
	for {
		input := readUserInputWithValidator("Enter a number: ",
			validateNumberOrDone, "Sorry. That's not a valid input.")
		if input == "done" {
			break
		} else {
			val, _ := strconv.Atoi(input)
			nums = append(nums, val)
		}
	}
	if len(nums) > 0 {
		dispNums(nums)
		fmt.Printf("The average is %.0f\n", average(nums))
		fmt.Printf("The minimum is %d\n", minimum(nums))
		fmt.Printf("The maximum is %d\n", maximum(nums))
		fmt.Printf("The standard deviation is %.2f\n", standardDiviation(nums))
	}
}

File: 36_集約/python/main.py
#!/usr/bin/env python
# coding: -utf8

import user_input as ui
import math
import functools


def validate_number_or_done(s):
    if s == "done":
        return s

    int(s)  # 例外がでないか試す.
    return s


def disp_nums(nums):
    print("Numbers: ", end="")
    sep = ""
    for i in nums:
        print(sep,  i, end="")
        sep = ","

    print("")


def average(nums):
    return sum(nums) / len(nums)


def standard_diviation(nums):
    ave = average(nums)
    tmp_sum = functools.reduce(lambda acc, e: acc + (ave - e) ** 2, nums, 0.0)
    variance = tmp_sum / len(nums)
    return math.sqrt(variance)


def round_pointn(d, n):
    multiplier = 1
    for i in range(n):
        multiplier *= 10
    d1 = d * multiplier
    d2 = round(d1)
    return d2 / multiplier


if __name__ == '__main__':
    nums = []
    while True:
        user_input = ui.read_user_input_with_validator(
            "Enter a number: ", validate_number_or_done,
            "Sorry. That's not a valid input.")
        if user_input == "done":
            break
        else:
            nums.append(int(user_input))

    if len(nums) != 0:
        disp_nums(nums)
        print(f"The average is {average(nums)}")
        print(f"The minimum is {min(nums)}")
        print(f"The maximum is {max(nums)}")
        print(
            f"The standard diviation is {round_pointn(standard_diviation(nums),2)}")
